jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/GracefulShutdownHandler.java

```diff
@@ -0,0 +1,161 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.server.handler;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpHeaderValue;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.component.Graceful;
+import org.eclipse.jetty.util.statistic.CounterStatistic;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handler to track active requests and allow them to gracefully complete.
+ */
+public class GracefulShutdownHandler extends Handler.Wrapper implements Graceful
+{
+    private static final Logger LOG = LoggerFactory.getLogger(GracefulShutdownHandler.class);
+
+    private CounterStatistic dispatchedStats = new CounterStatistic();
+    private Shutdown shutdown;
+
+    class ShutdownTrackingCallback extends Callback.Nested
+    {
+        final Request request;
+        final Response response;
+
+        public ShutdownTrackingCallback(Request request, Response response, Callback callback)
+        {
+            super(callback);
+            this.request = request;
+            this.response = response;
+            dispatchedStats.increment();
+        }
+
+        public void decrement()
+        {
+            dispatchedStats.decrement();
+        }
+
+        @Override
+        public void succeeded()
+        {
+            decrement();
+            super.succeeded();
+            if (isShutdown())
+            {
+                shutdown.check();
+            }
+        }
+
+        @Override
+        public void failed(Throwable x)
+        {
+            decrement();
+
+            Callback callback = Callback.from(getCallback(), shutdown::check);
+
+            if (isShutdown())
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Service Unavailable: {}", request.getHttpURI(), x);
+                Response.writeError(request, response, callback, HttpStatus.SERVICE_UNAVAILABLE_503);
```

Comment:
I'm questioning this now.  If an in progress request completes with an exception, should we not just write the error response normally to report that problem and decrement the count?   No need to swap in a 503 response... that is for any subsequent requests that arrive after shutdown.

Reply:
A few scenarios ...

Note: "When" can be ...

* Immediate : it happens immediately.
* Timeout : when the connection IdleTimeout or Shutdown Timeout occurs.
* Graceful : during Graceful Shutdown, not yet a Timeout.

Scenario | When | Result
---- | ---- | ---
A new request (entering `process()` after shutdown) | Immediate | a rapid "503 Service Unavailable" response.
An active request (`process()` returned true), but uncommitted | Timeout | "503 Service Unavailable" response.
An active request (`process()` not returned), but uncommitted | Timeout | "503 Service Unavailable" response.
An active request (`process()` returned true), committed | Timeout | Close connection
An active request (`process()` not returned), committed | Timeout | Close connection
An active request, using `writeError()` | Graceful | Don't change the error status
An active request, throwing uncaught Exception | Graceful | Use `writeError(Throwable)`
An active request, completing normally | Graceful | Decrement (do nothing extra)

Am I missing anything?